### PR TITLE
feat(worker/monitor): add metrics collector

### PR DESCRIPTION
**Pull request by YuriRDev**

### CODE REVIEW

File: ctl/Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
-[package]
-name = "ctl"
-version.workspace = true
-edition.workspace = true
-lints.workspace = true
-
-[dependencies]
+[package]
```
Comment (lffg) [inline ctl/Cargo.toml]: Please [configure](https://stackoverflow.com/a/13154031/7445826) your Git to use LF instead of CRLF. Note that commits unnecessarily changed entire unrelated files, such as this one.

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,4 @@
 target
 *.log
 ignore*
-.DS_Store
+.DS_Store
```
Comment (lffg) [inline .gitignore]: Unnecessary change, let's keep the convention of having a blank new line after each file. VSCode can configured to do that through `insertFinalNewline`.

File: worker/Cargo.toml [hunk 3]
```diff
@@ -5,3 +5,4 @@ edition.workspace = true
 lints.workspace = true
 
 [dependencies]
+sysinfo = "0.30.7"
```
Comment (lffg) [inline worker/Cargo.toml]: Use a workspace dependency instead so that all versions are centrally specified. [Learn more here.](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table)

File: worker/src/main.rs [hunk 4]
```diff
@@ -1,3 +1,9 @@
-fn main() {
-    println!("Hello, world!");
-}
+use sysinfo::System;
+mod metrics;
+use crate::metrics::{Metric, MemoryMeasure};
```
Comment (lffg) [inline worker/src/main.rs]:
```suggestion

use crate::metrics::{Metric, MemoryMeasure};

mod metrics;
```

Sorry, more nitpicks. I'll configure rustmft later so that this is enforced automatically by the formatter.

File: worker/src/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
+use sysinfo::System;
+
+pub enum MemoryMeasure {
+    Megabyte,
+    Gigabyte,
+}
+
+#[derive(Debug)]
+pub struct Metric {
```
Comment (lffg) [inline worker/src/metrics.rs]: I am not sure about this name, since _each_ one of this struct's field _is_ a metric. Maybe `MetricsReport`, `MetricBag`, or something like that?

File: worker/src/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+use sysinfo::System;
+
+pub enum MemoryMeasure {
+    Megabyte,
+    Gigabyte,
+}
```
Comment (lffg) [inline worker/src/metrics.rs]:
```suggestion
pub enum SpaceUnit {
    MiB,
    GiB,
}

```

File: worker/src/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            free_memory: get_memory(measure), 
+            cpu_usage: 10.0
+         }
+    }
+}
+
+fn get_memory(measure: MemoryMeasure) -> f64 {
+    match measure {
+        MemoryMeasure::Gigabyte => get_memory_as_gigabyte(),
+        MemoryMeasure::Megabyte => get_memory_as_megabyte(),
```
Comment (lffg) [inline worker/src/metrics.rs]: Hmm, you could unify these functions here by creating an associated method on the `SpaceUnit` (aka `MemoryMeasure`) enum that returns the multiplicative factor that converts bytes into the desired unit.

For example, `MiB` could return `1 / (1024 ** 2)` and `GiB` could return `1 / (1024 ** 3)`. And then, `(system.total_memory() - system.used_memory()) * unit.byte_conv_factor()`.

**Review by lffg — COMMENTED** [review]
Just some nitpicks. :P

File: worker/src/metrics.rs (hunk 6 above)
Comment (lffg) [reply worker/src/metrics.rs]: Also, I'm not sure about the "space" unit thing, since by "space" we usually mean something like meter, inches, etc. Maybe `BitUnit`, or `InformationUnit`? idk...

https://en.wikipedia.org/wiki/Byte

File: worker/src/metrics.rs (hunk 5 above)
Comment (gdiasag) [reply worker/src/metrics.rs]: While writing about the project's structure, I was skeptical about using the word "metric". I found it too formal and not intuitive enough. My suggestion would be something along the lines of this [kubernetes](https://github.com/kubernetes/kubernetes/blob/1f3c65c5f4c79a57427297d99985ab1493db7147/pkg/volume/util/fs/fs.go#L45) type definition, maybe `WorkerInfo`, `UsageReport`? Either way, `MetricsReport` also fits. 

image

File: worker/Cargo.toml (hunk 3 above)
Comment (lemosep) [reply worker/Cargo.toml]: When importing a unpublished crate (e.g. proto), is there an appropriate convention? I´ve seen the one used in the Cargo book [here](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies), is there a more concise way of doing it?

Comment (lffg) [reply worker/Cargo.toml]: Yes, after #5 is merged you'll be able to do `proto.workspace = true` on the `worker`'s Cargo.toml manifest. The path is being declared on the workspace's manifest:

https://github.com/esfericos/tucano/blob/829875a7b0f6352f8608f99ba02618eb22932e93/Cargo.toml#L9-L23

**Comment by YuriRDev** [comment]
Please, remove all `println!` on the source code.

File: worker/Cargo.toml [hunk 8]
```diff
@@ -7,3 +7,10 @@ edition.workspace = true
 workspace = true
 
 [dependencies]
+eyre = { workspace = true }
```
Comment (lffg) [inline worker/Cargo.toml]:
```suggestion
eyre = { workspace = true }
eyre.workspace = true
```

File: worker/src/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+use proto::common::node::Metrics;
+use sysinfo::System;
+
+const CPU_DELAY_IN_MILLIS: u64 = 500;
+
+/// Struct containing system's information such as CPU and Memory.
+pub struct MetricsReport {
+    system: System,
+}
+#[allow(clippy::cast_precision_loss)]
```
Comment (lffg) [inline worker/src/metrics.rs]: Move this to workspace root's cargo.toml

File: worker/src/main.rs [hunk 10]
```diff
@@ -1,3 +1,24 @@
...
+/// `clap` crate report interval placeholder
+/// The value 500 is for quick visualization purposes.
+const REPORT_INTERVAL_IN_MILLIS: u64 = 500;
+
+/// `clap` crate url placeholder
+const AGT_MGR_REQUEST_URL: &str = "http://localhost:8080/http/agt_mgr";
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mut monitor = MetricsSender::new(AGT_MGR_REQUEST_URL)?;
```
Comment (gdiasag) [inline worker/src/main.rs]: Not a monitor

```suggestion
    let mut metrics_sender = MetricsSender::new(AGT_MGR_REQUEST_URL)?;
```

File: worker/src/monitor/metrics_collector.rs [hunk 11]
```diff
@@ -0,0 +1,67 @@
+use std::{thread::sleep, time::Duration};
+
+use proto::common::node::Metrics;
+use sysinfo::System;
+
+const CPU_DELAY_IN_MILLIS: u64 = 500;
+
+/// Struct containing system's information such as CPU and Memory.
```
Comment (gdiasag) [inline worker/src/monitor/metrics_collector.rs]: The struct that actually contains that information is `sysinfo::System`, `MetricsCollector` is only a wrapper around it

File: proto/src/common/node.rs [hunk 12]
```diff
@@ -22,7 +22,8 @@ pub enum NodeKind {
 
 #[derive(Debug, Serialize, Deserialize)]
 pub struct Metrics {
-    pub cpu_usage: f64,
+    /// Collection of CPU usages.
+    pub cpu_usage: Vec<f64>,
```
Comment (gdiasag) [inline proto/src/common/node.rs]: I would just like to dive deeper in discussion about the more appropriate way of representing various CPU usages from a single node. Using a centralized representation such as percentage or total is certainly not accurate, but is it okay to use a `Vec` of them? I would argue that's valid but I'm not really sure.

File: worker/src/monitor/metrics_sender.rs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+/// A connection pool that sends the collected `Metrics` to the `http`component.
+pub struct MetricsSender {
+    client: Client,
+    base_url: String,
+}
+
+impl MetricsSender {
+    /// Instantiates a new `MetricsSender`.
+    ///
+    ///  Sends requests containing a JSON with the collected metrics
```
Comment (gdiasag) [inline worker/src/monitor/metrics_sender.rs]: The description of what is done by `MetricsSender` should be placed as a documentation of the struct definition instead of its constructor's.

File: worker/src/monitor/metrics_collector.rs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+pub struct MetricsCollector {
+    system: System,
+}
+impl MetricsCollector {
+    /// Instantiates a new `MetricsCollector`.
+    ///
+    /// Responsable for collecting metrics about the System, such as
```
Comment (gdiasag) [inline worker/src/monitor/metrics_collector.rs]: See note in `metrics_sender.rs`

File: worker/src/monitor/metrics_collector.rs [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    /// `CPU_DELAY_IN_MILLIS`.
+    pub fn get_metrics(&mut self) -> Metrics {
+        Metrics {
+            cpu_usage: self.get_cpu_usage(),
+            mem_total_mib: self.get_total_memory(),
+            mem_used_mib: self.get_used_memory(),
+        }
+    }
+
+    fn get_cpu_usage(&mut self) -> Vec<f64> {
```
Comment (gdiasag) [inline worker/src/monitor/metrics_collector.rs]: Does not return a single usage value 

```suggestion
    fn get_cpus_usage(&mut self) -> Vec<f64> {
```

File: worker/src/monitor/metrics_collector.rs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        self.system.refresh_cpu();
+        sleep(Duration::from_millis(CPU_DELAY_IN_MILLIS));
+        self.system.refresh_cpu();
+
+        let all_cpus_usages: Vec<f64> = self
```
Comment (gdiasag) [inline worker/src/monitor/metrics_collector.rs]:
```suggestion
        let cpus_usage: Vec<f64> = self
```

File: proto/src/common/node.rs (hunk 12 above)
Comment (lffg) [reply proto/src/common/node.rs]: I'd also prefer a f64 as a starting point... Using (and sending) a vec seems really overkill for what we're proposing, even though I can see the motivations behind it.

File: worker/src/monitor/metrics_collector.rs (hunk 11 above)
Comment (lffg) [reply worker/src/monitor/metrics_collector.rs]: Also it doesn't contain _the information_, but it can be used as a _mechanism to get_ the up-do-date information.

Comment (lffg) [reply worker/src/monitor/metrics_collector.rs]: Yeah, naming is hard.

File: proto/src/common/node.rs (hunk 12 above)
Comment (lemosep) [reply proto/src/common/node.rs]: I think we should stick with the `Vec`, representing the CPUs usages as a single f64 is *very* inaccurate and could lead to performance issues such as ignoring `workers` that are capable of running a task just because one of it's CPUs is being highly used. WDYT @gustavodiasag  @lffg?

Comment (lemosep) [reply proto/src/common/node.rs]: Tbh, for the sake of closing this PR so we can move on to other components, let's keep it simple by now and just use a `f64`, let time tell if that `Vec` will be a real need.

File: proto/src/common/node.rs [hunk 17]
```diff
@@ -22,6 +22,7 @@ pub enum NodeKind {
 
 #[derive(Debug, Serialize, Deserialize)]
 pub struct Metrics {
+    /// The CPU usage, in MiB.
```
Comment (lffg) [inline proto/src/common/node.rs]:
```suggestion
    /// The average CPU usage.
```

File: Cargo.toml [hunk 18]
```diff
@@ -18,12 +18,17 @@ chrono = { version = "0.4", default-features = false, features = [
     "std",
     "serde",
 ] }
+eyre = "0.6.12"
+reqwest = { version = "0.12.1" , features = ["json"] }
 serde = { version = "1", features = ["derive"] }
 serde_json = "1"
+sysinfo = "0.30.7"
+tokio = { version = "1.36.0", features = ["full"] }
```
Comment (lffg) [inline Cargo.toml]:
```suggestion
tokio = "1.36"
```

Please only pick the features you'll need. (see my next comment)

File: worker/Cargo.toml [hunk 19]
```diff
@@ -7,3 +7,10 @@ edition.workspace = true
 workspace = true
 
 [dependencies]
+eyre.workspace = true
+proto.workspace = true
+reqwest.workspace = true
+serde.workspace = true
+serde_json.workspace = true
+sysinfo.workspace = true
+tokio.workspace = true
```
Comment (lffg) [inline worker/Cargo.toml]:
```suggestion
tokio = { workspace = true, features = ["rt-multi-thread", "macros"] }
```

File: worker/src/main.rs [hunk 20]
```diff
@@ -1,3 +1,24 @@
...
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mut metrics_sender = MetricsSender::new(AGT_MGR_REQUEST_URL)?;
+    let mut metrics_report: MetricsCollector = MetricsCollector::new();
+
+    loop {
+        let metrics = metrics_report.get_metrics();
+        metrics_sender.send_request(&metrics).await?;
+        sleep(Duration::from_millis(REPORT_INTERVAL_IN_MILLIS));
```
Comment (lffg) [inline worker/src/main.rs]: You should use `tokio`'s `sleep` here.

File: worker/src/monitor/metrics_collector.rs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+        self.system.refresh_cpu();
+        sleep(Duration::from_millis(CPU_DELAY_IN_MILLIS));
+        self.system.refresh_cpu();
+
+        let cpus_usages: Vec<f64> = self
+            .system
+            .cpus()
+            .iter()
+            .map(|cpu| f64::from(cpu.cpu_usage()))
+            .collect();
```
Comment (lffg) [inline worker/src/monitor/metrics_collector.rs]: You can call `sum` directly here to avoid the vector allocation :)

File: worker/Cargo.toml [hunk 22]
```diff
@@ -7,3 +7,10 @@ edition.workspace = true
 workspace = true
 
 [dependencies]
+eyre.workspace = true
+proto.workspace = true
+reqwest.workspace = true
+serde.workspace = true
+serde_json.workspace = true
```
Comment (lffg) [inline worker/Cargo.toml]: I don't think this dependency (serde_json) is necessary here.

File: worker/src/monitor/metrics_collector.rs [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+        Metrics {
+            cpu_usage: self.get_cpu_usage(),
+            mem_total_mib: self.get_total_memory(),
+            mem_used_mib: self.get_used_memory(),
+        }
+    }
+
+    fn get_cpu_usage(&mut self) -> f64 {
+        self.system.refresh_cpu();
+        sleep(Duration::from_millis(CPU_DELAY_IN_MILLIS));
```
Comment (lffg) [inline worker/src/monitor/metrics_collector.rs]: As we're using async, we should not use a sync sleep here, as it would block the thread thus preventing other tasks from making progress. That's _not_ currently a problem per se, but it could be in the future--in general, we should always avoid making blocking calls from async code.

We have two options:
1. Make this function (and all of its callers) `async` so that we may call tokio's `sleep` here. In this case, we'd also need to parameterize the delay since it's going to be dynamic from the command-line arguments.
2. Remove the sleep from this function by delegating it to the caller. In this case, the caller is `main`, which _should_ only call `get_metrics` after specific intervals of time.

Personally, I'm more fond of the latter option.

Comment (gdiasag) [reply worker/src/monitor/metrics_collector.rs]: I also vouch for option 2 here

Comment (lemosep) [reply worker/src/monitor/metrics_collector.rs]: But how to make main call `get_metrics` within these intervals without a `loop` + `sleep` combination? How would that be structured?

File: worker/src/monitor/metrics_collector.rs (hunk 21 above)
Comment (lemosep) [reply worker/src/monitor/metrics_collector.rs]: True! I need the length of the `Vec` before reducing it to a single value though. So I came up with this:: 
```rs
        let amount_cpus: usize = self
            .system
            .cpus()
            .len();

        let sum_cpus_usages: f64 = self
            .system
            .cpus()
            .iter()
            .map(|cpu| f64::from(cpu.cpu_usage()))
            .sum(); 

        sum_cpus_usages / amount_cpus as f64
 ```

File: worker/src/monitor/metrics_collector.rs (hunk 23 above)
Comment (lffg) [reply worker/src/monitor/metrics_collector.rs]: The loop and a sleep (from tokio) would still exist. The point is that they wouldn’t be on the colletor module, but rather on one of its calling modules.

**Review by lemosep — APPROVED** [review]
LGTM!

**Review by lffg — APPROVED** [review]
Good work, guys!!
